### PR TITLE
Upgrade Jet Factories to Multifactories

### PR DESCRIPTION
**Pull request by ruse-traveler**
### Briefly, what does this PR introduce?

This PR upgrades the existing jet factories to multifactories and also modifies the jet reconstruction algorithm such that it returns a PODIO collection of reconstructed particles rather than a standard vector.

### What kind of change does this PR introduce?
- [ ] Bug fix (issue #__)
- [ X] New feature (issue #__)
- [ ] Documentation update
- [ ] Other: __

### Please check if this PR fulfills the following:
- [ ] Tests for the changes have been added
- [ ] Documentation has been added / updated
- [X ] Changes have been communicated to collaborators

### Does this PR introduce breaking changes? What changes might users need to make to their code?

No.

### Does this PR change default behavior?

No.

### CODE REVIEW

File: src/algorithms/reco/JetReconstruction.cc [hunk 1]
```diff
@@ -23,31 +23,32 @@ namespace eicrecon {
     m_log->trace("Initialized");
   }
 
-  std::vector<edm4eic::ReconstructedParticle*> JetReconstruction::execute(
+  edm4eic::ReconstructedParticleCollection* JetReconstruction::execute(
```
Comment (veprbl) [inline src/algorithms/reco/JetReconstruction.cc:26]:
```suggestion
  std::unique_ptr<edm4eic::ReconstructedParticleCollection> JetReconstruction::execute(
```

File: src/algorithms/reco/JetReconstruction.cc [hunk 2]
```diff
@@ -23,31 +23,32 @@ namespace eicrecon {
...
     m_log->trace("  Number of particles: {}", momenta.size());
 
     // Particles for jet reconstrution
     std::vector<PseudoJet> particles;
     for (const auto &mom : momenta) {
+
       // Only cluster particles within the given pt Range
       if ((mom->pt() > m_minCstPt) && (mom->pt() < m_maxCstPt)) {
-        particles.emplace_back(mom->px(), mom->py(), mom->pz(), mom->e() );
+        particles.push_back( PseudoJet(mom->px(), mom->py(), mom->pz(), mom->e()) );
```
Comment (wdconinc) [inline src/algorithms/reco/JetReconstruction.cc:46]: You are reverting changes in the target branch.

**Review by wdconinc — CHANGES_REQUESTED** [review]

File: src/algorithms/reco/JetReconstruction.cc [hunk 3]
```diff
@@ -57,41 +58,48 @@ namespace eicrecon {
...
-        cst_edm.setEnergy(csts[j].e());
-        cst_edm.setMass(csts[j].m());
-        //jet_edm.addToParticles(cst_edm);  // FIXME: global issue with podio reference
+        edm4eic::MutableReconstructedParticle cst_output;
+        cst_output.setType(1);
+        cst_output.setPDG(i);
+        cst_output.setMomentum(edm4hep::Vector3f(csts[j].px(), csts[j].py(), csts[j].pz()));
+        cst_output.setEnergy(csts[j].e());
+        cst_output.setMass(csts[j].m());
+        //jet_output.addToParticles(cst_output);  // FIXME: global issue with podio reference
```
Comment (veprbl) [inline src/algorithms/reco/JetReconstruction.cc:90]: This one would need to be routed properly, something along the lines of
```c++
add_to_particle(jet_output, input_collection[csts[j].user_index()])
```
where
```c++
template <typename T>
add_to_particle(edm4eic::MutableReconstructedParticle &p, T &other);

template <>
add_to_particle(edm4eic::MutableReconstructedParticle &p, edm4eic::ReconstructedParticle &other) {
  p.addToParticles(other);
}

template <>
add_to_particle(edm4eic::MutableReconstructedParticle &p, edm4eic::MCParticle &other) {
  p.addToMCParticles(other);
}
```

File: src/algorithms/reco/JetReconstruction.h [hunk 4]
```diff
@@ -6,11 +6,13 @@
 
 // standard c includes
 #include <vector>
-// DD4hep includes
+// jana includes
```
Comment (veprbl) [inline src/algorithms/reco/JetReconstruction.h:9]:
```suggestion
```

File: src/algorithms/reco/JetReconstruction.cc [hunk 5]
```diff
@@ -23,31 +23,32 @@ namespace eicrecon {
     m_log->trace("Initialized");
   }
 
-  std::vector<edm4eic::ReconstructedParticle*> JetReconstruction::execute(
+  edm4eic::ReconstructedParticleCollection* JetReconstruction::execute(
     const std::vector<const edm4hep::LorentzVectorE*> momenta) {
```
Comment (veprbl) [inline src/algorithms/reco/JetReconstruction.cc:27]: Would need to use templated type here
```suggestion
    const T &input_collection) {
```

**Review by veprbl — COMMENTED** [review]
Looks good. Mostly rambling about unusual formatting.
Please use smart pointers. We also would like to get proper associations, I'm putting down some ideas on how to integrate that.

File: src/global/reco/reco.cc [hunk 6]
```diff
@@ -37,7 +36,6 @@ void InitPlugin(JApplication *app) {
     app->Add(new JChainMultifactoryGeneratorT<MatchClusters_factory>(
         "ReconstructedParticlesWithAssoc",
         { "EcalEndcapNClusters",
-          "EcalBarrelScFiClusters",
```
Comment (veprbl) [inline src/global/reco/reco.cc:40]: This doesn't seem right...

File: src/global/reco/reco.cc [hunk 7]
```diff
@@ -65,22 +63,17 @@ void InitPlugin(JApplication *app) {
...
-        "EcalEndcapNClusterAssociations",
-        "EcalEndcapPClusterAssociations",
-        "EcalEndcapPInsertClusterAssociations",
-        "EcalLumiSpecClusterAssociations",
-        },
-        "ReconstructedElectrons"
+    app->Add(new JChainMultifactoryGeneratorT<GeneratedJets_factory>(
+             "GeneratedJets",
+            {"MCParticles"},
+            {"GeneratedJets"},
```
Comment (veprbl) [inline src/global/reco/reco.cc:69]: We don't need to change these factories to mutifactories at all. They need to output jets (currently modeled as `edm4hep::ReconstructedParticles`, we then refer to constituents that are already written in the MCParticles and ReconstructedParticles collections.

File: src/global/reco/reco.cc [hunk 8]
```diff
@@ -65,22 +63,17 @@ void InitPlugin(JApplication *app) {
...
+            app
+    ));
+    app->Add(new JChainMultifactoryGeneratorT<ReconstructedJets_factory>(
+            "ReconstrucctedJets",
+            {"ReconstructedParticles"},
```
Comment (bspage912) [inline src/global/reco/reco.cc:74]: Change input collection to ReconstructedChargedParticles?

File: src/algorithms/reco/JetReconstruction.cc [hunk 9]
```diff
@@ -23,31 +23,32 @@ namespace eicrecon {
...
     }
 
     m_log->trace("  Number of particles: {}", momenta.size());
 
     // Particles for jet reconstrution
     std::vector<PseudoJet> particles;
     for (const auto &mom : momenta) {
+
       // Only cluster particles within the given pt Range
       if ((mom->pt() > m_minCstPt) && (mom->pt() < m_maxCstPt)) {
```
Comment (bspage912) [inline src/algorithms/reco/JetReconstruction.cc:45]: May be for another PR to address, but logically, I'm not sure we want to impose the particle / track pT cuts here as we likely want them to be different depending on the input. We usually place some threshold on tracks (and calorimeter clusters) based on the tracker performance, but place a much lower pT threshold on stable MC particles.

**Review by bspage912 — COMMENTED** [review]
I have no comment on the code changes themselves as I am still trying to understand all this myself :-). 

I added comments on using ReconstructedChargedParticles vs ReconstructedParticles as the input in the ReconstructedJets_factory, on possibly restricting the input particle species in GeneratedJets_factory, and on where we logically want to place input track/particle pT thresholds.

File: src/algorithms/reco/JetReconstruction.cc (hunk 1 above)
Comment (ruse-traveler) [reply src/algorithms/reco/JetReconstruction.cc:26]: Whoops! Thank you for catching that: I deleted the wrong line while cleaning up.

File: src/algorithms/reco/JetReconstruction.cc (hunk 2 above)
Comment (ruse-traveler) [reply src/algorithms/reco/JetReconstruction.cc:46]: Ah! Thank you for catching that!

File: src/global/reco/reco.cc (hunk 6 above)
Comment (ruse-traveler) [reply src/global/reco/reco.cc:40]: 😩 Aggh. I deleted that by accident. Thank you for catching that!

File: src/global/reco/reco.cc (hunk 7 above)
Comment (ruse-traveler) [reply src/global/reco/reco.cc:69]: Out of curiosity: if we wanted to have multiple jet collections (e.g. a collection of track-only jets and then a collection of particle flow jets), would it be ideal to use multifactories?

File: src/global/reco/ReconstructedJets_factory.cc [hunk 10]
```diff
@@ -1,49 +1,74 @@
...
         // Nothing to do here
-    }
+        return;
+
+    }  // end 'BeginRun(std::shared_ptr<JEvent&>)'
+
+
 
     void ReconstructedJets_factory::Process(const std::shared_ptr<const JEvent> &event) {
-        auto rc_particles = event->Get<edm4eic::ReconstructedParticle>("ReconstructedParticles");
```
[the remark this answers is not in the document]
Comment (ruse-traveler) [reply src/global/reco/ReconstructedJets_factory.cc:33]: Whoops! Good catch: that should be ReconstructedChargedParticles. But that's correct! For multifactories, the input and output tags are specified in the reco.cc file.

File: src/global/reco/GeneratedJets_factory.cc [hunk 11]
```diff
@@ -1,52 +1,77 @@
...
+          std::vector<const edm4hep::LorentzVectorE*> momenta;
+          for (const auto& particle : input) {
+
+            // select only final state particles
+            const bool is_final_state = (particle -> getGeneratorStatus() == 1);
+            if (!is_final_state) continue;
+
+            const auto& momentum = particle -> getMomentum();
+            const auto& energy   = particle -> getEnergy();
+            momenta.push_back(new edm4hep::LorentzVectorE(momentum.x, momentum.y, momentum.z, energy));
```
[the remark this answers is not in the document]
Comment (ruse-traveler) [reply src/global/reco/GeneratedJets_factory.cc:59]: That's a good point... And I would say we should. But what if we add an option to the algorithm to pick either charged or all generated particles?

File: src/algorithms/reco/JetReconstruction.cc (hunk 9 above)
Comment (ruse-traveler) [reply src/algorithms/reco/JetReconstruction.cc:45]: Definitely! I was thinking on addressing this in another PR, but in making the options for both factories configurable, we'll also be able to set the kinematic thresholds differently between the reconstructed and generated jets.

File: src/global/reco/reco.cc (hunk 7 above)
Comment (ruse-traveler) [reply src/global/reco/reco.cc:69]: Ahhhh I see. I guess for my edification: are multifactories exclusively for collections that require multiple inputs (e.g. reconstructed particles)?

Comment (veprbl) [reply src/global/reco/reco.cc:69]: Multifactories are needed for factories that want to produce multiple outputs.

Just to clarify (cc @wdconinc), we are likely to convert all factories (even those producing a single output) to multfactories eventually. Presumably that would not impact the performance, but a consistent interface would simplify implementation of Gaudi-style steering via configuration in Python. Latter is the current informal consensus regarding the direction to go forward. Having said that, producing multiple kinds of jets from the same factories doesn't seem like a sound choice: there is no shared computation between the two procedures, but the shared state opens them up to introducing bugs. We would be much better off making algorithms sufficiently configurable and running those from separately defined factories. In that sense, you might be doing the right thing for the wrong reason.

File: src/global/reco/reco.cc [hunk 12]
```diff
@@ -65,22 +63,17 @@ void InitPlugin(JApplication *app) {
...
-        },
-        "ReconstructedElectrons"
+    app->Add(new JChainMultifactoryGeneratorT<GeneratedJets_factory>(
+             "GeneratedJets",
+            {"MCParticles"},
+            {"GeneratedJets"},
+            app
+    ));
+    app->Add(new JChainMultifactoryGeneratorT<ReconstructedJets_factory>(
+            "ReconstrucctedJets",
```
Comment (veprbl) [inline src/global/reco/reco.cc:73]:
```suggestion
            "ReconstructedJets",
```

File: src/global/reco/ReconstructedJets_factory.h [hunk 13]
```diff
@@ -3,37 +3,49 @@
 
 #pragma once
 
-#include <edm4eic/ReconstructedParticleCollection.h>
+// jana includes
```
Comment (veprbl) [inline src/global/reco/ReconstructedJets_factory.h:6]:
```suggestion
```

File: src/global/reco/GeneratedJets_factory.cc (hunk 11 above)
[the remark this answers is not in the document]
Comment (ruse-traveler) [reply src/global/reco/GeneratedJets_factory.cc:57]: Forgot to comment here: but to implement this, I added an additional condition to the generated jets factory to pick out only charged particles!

**Comment by ruse-traveler** [comment]
Hi all, just wanted to touch base on this PR to make sure it gets in before the September campaign. If everyone's okay with implementing PODIO references in a separate PR, then I think I've addressed all comments so far. (If no one objects, then I'll mark all the above conversations as resolved.)

Let me know if I missed anything or if there's anything else I need to do to make sure this gets merged!

**Comment by wdconinc** [comment]
> Let me know if I missed anything or if there's anything else I need to do to make sure this gets merged!

There are conflicts...

**Comment by ruse-traveler** [comment]
> > Let me know if I missed anything or if there's anything else I need to do to make sure this gets merged!
> 
> There are conflicts...

Whoops! I should've specified: aside from the conflicts! 😓 

(I'll take care of those, though!)

File: src/global/reco/GeneratedJets_factory.cc [hunk 14]
```diff
@@ -1,52 +1,72 @@
...
-        std::vector<const edm4hep::LorentzVectorE*> momenta;
-        for (const auto& p : mc_particles) {
-          if (p -> getGeneratorStatus() == 1) {
-            const auto& mom    = p -> getMomentum();
-            const auto& energy = p -> getEnergy();
-            momenta.push_back(new edm4hep::LorentzVectorE(mom.x, mom.y, mom.z, energy));
+
+        // loop over input tags
+        size_t iInput = 0;
+        for (const std::string& input_tag : GetInputTags()) {
```
Comment (veprbl) [inline src/global/reco/GeneratedJets_factory.cc:41]: Why are you iterating over input collections? This can't work because down below you overwrite previous result with the latest collection.

I would stick to just one input particle collection `GetInputTags()[0]` and leave the other inputs reserved for possible future use. If we need to merge multiple particle collections, let's make that code into a separate factory.

Comment (ruse-traveler) [reply src/global/reco/GeneratedJets_factory.cc:41]: Ah! This is a holdover from my misunderstanding of what a multifactory was for: my thinking was that I would loop over specified input collections and produce an output for each one. I didn't remove it in the last batch of commits just because I thought it wouldn't make a difference.

But I agree with your suggestion: I'll stick to `GetTags()[0]` and reserve the others for future use!

File: src/algorithms/reco/JetReconstruction.cc [hunk 15]
```diff
@@ -23,26 +23,27 @@ namespace eicrecon {
...
-        particles.emplace_back(mom->px(), mom->py(), mom->pz(), mom->e() );
+        particles.emplace_back( PseudoJet(mom->px(), mom->py(), mom->pz(), mom->e()) );
```
Comment (veprbl) [inline src/algorithms/reco/JetReconstruction.cc:46]:
```suggestion
        particles.emplace_back(mom->px(), mom->py(), mom->pz(), mom->e());
```

File: src/global/reco/reco.cc [hunk 16]
```diff
@@ -34,7 +33,7 @@ void InitPlugin(JApplication *app) {
     app->Add(new JChainMultifactoryGeneratorT<MatchClusters_factory>(
         "ReconstructedParticlesWithAssoc",
         { "EcalEndcapNClusters",
-          "EcalBarrelScFiClusters",
+          "EcalBarrelSciFiClusters",
```
Comment (veprbl) [inline src/global/reco/reco.cc:36]:
```suggestion
          "EcalBarrelScFiClusters",
```

File: src/global/reco/reco.cc [hunk 17]
```diff
@@ -62,22 +61,17 @@ void InitPlugin(JApplication *app) {
...
-    app->Add(new JChainFactoryGeneratorT<ReconstructedElectrons_factory>(
-        {"MCParticles", "ReconstructedChargedParticles", "ReconstructedChargedParticleAssociations",
-        "EcalBarrelScFiClusterAssociations",
-        "EcalEndcapNClusterAssociations",
-        "EcalEndcapPClusterAssociations",
-        "EcalEndcapPInsertClusterAssociations",
-        "EcalLumiSpecClusterAssociations",
-        },
-        "ReconstructedElectrons"
+    app->Add(new JChainMultifactoryGeneratorT<GeneratedJets_factory>(
```
Comment (veprbl) [inline src/global/reco/reco.cc:64]: Submit a separate PR for this
```suggestion
    app->Add(new JChainFactoryGeneratorT<ReconstructedElectrons_factory>(
         {"MCParticles", "ReconstructedChargedParticles", "ReconstructedChargedParticleAssociations",
         "EcalBarrelScFiClusterAssociations",
         "EcalEndcapNClusterAssociations",
         "EcalEndcapPClusterAssociations",
         "EcalEndcapPInsertClusterAssociations",
         "EcalLumiSpecClusterAssociations",
         },
         "ReconstructedElectrons"

    app->Add(new JChainMultifactoryGeneratorT<GeneratedJets_factory>(
```

File: src/global/reco/reco.cc [hunk 18]
```diff
@@ -19,7 +19,6 @@
 #include "InclusiveKinematicsSigma_factory.h"
 #include "GeneratedJets_factory.h"
 #include "ReconstructedJets_factory.h"
-#include "ReconstructedElectrons_factory.h"
```
Comment (veprbl) [inline src/global/reco/reco.cc:22]:
```suggestion
#include "ReconstructedElectrons_factory.h"

```

File: src/global/reco/reco.cc [hunk 19]
```diff
@@ -62,22 +62,27 @@ void InitPlugin(JApplication *app) {
...
-        "ReconstructedElectrons"
+         {"MCParticles", "ReconstructedChargedParticles", "ReconstructedChargedParticleAssociations",
+         "EcalBarrelScFiClusterAssociations",
+         "EcalEndcapNClusterAssociations",
+         "EcalEndcapPClusterAssociations",
+         "EcalEndcapPInsertClusterAssociations",
+         "EcalLumiSpecClusterAssociations",
+         },
+         "ReconstructedElectrons"
+
```
Comment (veprbl) [inline src/global/reco/reco.cc:74]:
```suggestion
    ));

```

File: src/global/reco/GeneratedJets_factory.h [hunk 20]
```diff
@@ -3,37 +3,47 @@
...
             public SpdlogMixin {
 
     public:
-        explicit GeneratedJets_factory(std::vector<std::string> default_input_tags):
-            JChainFactoryT<edm4eic::ReconstructedParticle>(std::move(default_input_tags)) {
-        }
+
+        // ctor
+        explicit GeneratedJets_factory(std::string tag,
+                                       const std::vector<std::string>& input_tags,
```
Comment (veprbl) [inline src/global/reco/GeneratedJets_factory.h:26]:
```suggestion
        explicit GeneratedJets_factory(const std::vector<std::string>& input_tags,
```

File: src/global/reco/ReconstructedJets_factory.h [hunk 21]
```diff
@@ -3,37 +3,47 @@
...
             public SpdlogMixin {
 
     public:
-        explicit ReconstructedJets_factory(std::vector<std::string> default_input_tags):
-            JChainFactoryT<edm4eic::ReconstructedParticle>(std::move(default_input_tags)) {
-        }
+
+        // ctor
+        explicit ReconstructedJets_factory(std::string tag,
+                                           const std::vector<std::string>& input_tags,
```
Comment (veprbl) [inline src/global/reco/ReconstructedJets_factory.h:26]:
```suggestion
        explicit ReconstructedJets_factory(const std::vector<std::string>& input_tags,
```

File: src/global/reco/reco.cc [hunk 22]
```diff
@@ -62,22 +62,28 @@ void InitPlugin(JApplication *app) {
...
+         "EcalEndcapNClusterAssociations",
+         "EcalEndcapPClusterAssociations",
+         "EcalEndcapPInsertClusterAssociations",
+         "EcalLumiSpecClusterAssociations",
+         },
+         "ReconstructedElectrons"
     ));
 
+    app->Add(new JChainMultifactoryGeneratorT<GeneratedJets_factory>(
+             "GeneratedJets",
```
Comment (veprbl) [inline src/global/reco/reco.cc:77]:
```suggestion
```

File: src/global/reco/reco.cc [hunk 23]
```diff
@@ -62,22 +62,28 @@ void InitPlugin(JApplication *app) {
...
+            {"MCParticles"},
+            {"GeneratedJets"},
+            app
+    ));
+    app->Add(new JChainMultifactoryGeneratorT<ReconstructedJets_factory>(
+            "ReconstructedJets",
```
Comment (veprbl) [inline src/global/reco/reco.cc:83]:
```suggestion
```

File: src/global/reco/ReconstructedJets_factory.cc [hunk 24]
```diff
@@ -1,49 +1,63 @@
...
 
+        // grab input collection
+        auto input = event->Get<edm4eic::ReconstructedParticle>(GetInputTags()[0]);
```
Comment (veprbl) [inline src/global/reco/ReconstructedJets_factory.cc:40]: Avoid conversion to `std::vector`
```suggestion
        auto input = static_cast<const edm4eic::ReconstructedParticleCollection*>(event->GetCollectionBase(GetInputTags()[0]));
```

File: src/global/reco/GeneratedJets_factory.cc [hunk 25]
```diff
@@ -1,52 +1,67 @@
...
+
+    }  // end 'BeginRun(std::shared_ptr<JEvent&>)'
+
+
 
     void GeneratedJets_factory::Process(const std::shared_ptr<const JEvent> &event) {
-        auto mc_particles = event->Get<edm4hep::MCParticle>("MCParticles");
 
+        // grab input collection
+        auto input = event->Get<edm4hep::MCParticle>(GetInputTags()[0]);
```
Comment (veprbl) [inline src/global/reco/GeneratedJets_factory.cc:40]: Avoid conversion to `std::vector`
```suggestion
        auto input = static_cast<const edm4hep::MCParticleCollection*>(event->GetCollectionBase(GetInputTags()[0]));
```

File: src/global/reco/GeneratedJets_factory.h [hunk 26]
```diff
@@ -3,37 +3,47 @@
...
+        void BeginRun(const std::shared_ptr<const JEvent> &event) override;
 
         /** Event by event processing **/
         void Process(const std::shared_ptr<const JEvent> &event) override;
 
     protected:
-        JetReconstruction m_jet_algo;
 
-    };
+        std::vector<std::string> m_input_tags;
```
Comment (veprbl) [inline src/global/reco/GeneratedJets_factory.h:44]: Avoid adding dead code
```suggestion
```

File: src/global/reco/ReconstructedJets_factory.h [hunk 27]
```diff
@@ -3,37 +3,47 @@
...
+        void BeginRun(const std::shared_ptr<const JEvent> &event) override;
 
         /** Event by event processing **/
         void Process(const std::shared_ptr<const JEvent> &event) override;
 
     protected:
-        JetReconstruction m_jet_algo;
 
-    };
+        std::vector<std::string> m_input_tags;
```
Comment (veprbl) [inline src/global/reco/ReconstructedJets_factory.h:44]:
```suggestion
```

File: src/global/reco/reco.cc [hunk 28]
```diff
@@ -62,22 +62,28 @@ void InitPlugin(JApplication *app) {
...
-        },
-        "ReconstructedElectrons"
+         {"MCParticles", "ReconstructedChargedParticles", "ReconstructedChargedParticleAssociations",
+         "EcalBarrelScFiClusterAssociations",
+         "EcalEndcapNClusterAssociations",
+         "EcalEndcapPClusterAssociations",
+         "EcalEndcapPInsertClusterAssociations",
+         "EcalLumiSpecClusterAssociations",
+         },
+         "ReconstructedElectrons"
```
Comment (veprbl) [inline src/global/reco/reco.cc:73]:
```suggestion
        {"MCParticles", "ReconstructedChargedParticles", "ReconstructedChargedParticleAssociations",
        "EcalBarrelScFiClusterAssociations",
        "EcalEndcapNClusterAssociations",
        "EcalEndcapPClusterAssociations",
        "EcalEndcapPInsertClusterAssociations",
        "EcalLumiSpecClusterAssociations",
        },
        "ReconstructedElectrons"
```

File: src/global/reco/GeneratedJets_factory.cc [hunk 29]
```diff
@@ -1,52 +1,67 @@
...
-        }
+        for (const auto& particle : input) {
 
-        auto jets = m_jet_algo.execute(momenta);
-        for (const auto &mom : momenta) {
-          delete mom;
+            // select only final state charged particles
+            const bool is_final_state = (particle->getGeneratorStatus() == 1);
+            const bool is_charged = (particle->getCharge() != 0.);
+            if (!is_final_state || !is_charged) continue;
```
Comment (veprbl) [inline src/global/reco/GeneratedJets_factory.cc:49]: Doesn't fit the description - please submit separately
```suggestion
            if (!is_final_state) continue;
```

File: src/global/reco/reco.cc [hunk 30]
```diff
@@ -62,22 +62,28 @@ void InitPlugin(JApplication *app) {
...
 
+    app->Add(new JChainMultifactoryGeneratorT<GeneratedJets_factory>(
+             "GeneratedJets",
+            {"MCParticles"},
+            {"GeneratedJets"},
+            app
+    ));
+    app->Add(new JChainMultifactoryGeneratorT<ReconstructedJets_factory>(
+            "ReconstructedJets",
+            {"ReconstructedChargedParticles"},
```
Comment (veprbl) [inline src/global/reco/reco.cc:84]: Please submit separately
```suggestion
            {"ReconstructedParticles"},
```

File: src/global/reco/GeneratedJets_factory.cc [hunk 31]
```diff
@@ -1,52 +1,66 @@
...
-
+// standard c includes
 #include <memory>
-
 #include <JANA/JEvent.h>
-#include "services/log/Log_service.h"
-
+#include <services/log/Log_service.h>
+// event data model definitions
 #include <edm4hep/MCParticle.h>
```
Comment (veprbl) [inline src/global/reco/GeneratedJets_factory.cc:9]:
```suggestion
#include <edm4hep/MCParticleCollection.h>
```

File: src/global/reco/ReconstructedJets_factory.cc [hunk 32]
```diff
@@ -1,49 +1,63 @@
...
-#include "ReconstructedJets_factory.h"
-
+// standard c includes
 #include <memory>
-
 #include <JANA/JEvent.h>
-#include "services/log/Log_service.h"
+#include <services/log/Log_service.h>
+// event data model definitions
+#include <edm4eic/ReconstructedParticle.h>
```
Comment (veprbl) [inline src/global/reco/ReconstructedJets_factory.cc:9]:
```suggestion
#include <edm4eic/ReconstructedParticleCollection.h>
```

File: src/global/reco/ReconstructedJets_factory.cc [hunk 33]
```diff
@@ -1,49 +1,63 @@
...
-        for (const auto& p : rc_particles) {
-          // TODO: Need to exclude the scattered electron
-          const auto& mom    = p -> getMomentum();
-          const auto& energy = p -> getEnergy();
-          momenta.push_back(new edm4hep::LorentzVectorE(mom.x, mom.y, mom.z, energy));
+        for (const auto& particle : input) {
+
+            // TODO: Need to exclude the scattered electron
+            const auto& momentum = particle->getMomentum();
+            const auto& energy = particle->getEnergy();
```
Comment (veprbl) [inline src/global/reco/ReconstructedJets_factory.cc:48]:
```suggestion
            const auto& momentum = particle.getMomentum();
            const auto& energy = particle.getEnergy();
```

File: src/global/reco/GeneratedJets_factory.cc [hunk 34]
```diff
@@ -1,52 +1,66 @@
...
 
-        auto jets = m_jet_algo.execute(momenta);
-        for (const auto &mom : momenta) {
-          delete mom;
+            // select only final state charged particles
+            const bool is_final_state = (particle->getGeneratorStatus() == 1);
+            if (!is_final_state) continue;
+
+            const auto& momentum = particle->getMomentum();
+            const auto& energy = particle->getEnergy();
```
Comment (veprbl) [inline src/global/reco/GeneratedJets_factory.cc:51]:
```suggestion
            const auto& momentum = particle.getMomentum();
            const auto& energy = particle.getEnergy();
```

File: src/global/reco/GeneratedJets_factory.cc [hunk 35]
```diff
@@ -1,52 +1,66 @@
...
-            momenta.push_back(new edm4hep::LorentzVectorE(mom.x, mom.y, mom.z, energy));
-          }
-        }
+        for (const auto& particle : input) {
 
-        auto jets = m_jet_algo.execute(momenta);
-        for (const auto &mom : momenta) {
-          delete mom;
+            // select only final state charged particles
+            const bool is_final_state = (particle->getGeneratorStatus() == 1);
```
Comment (veprbl) [inline src/global/reco/GeneratedJets_factory.cc:47]:
```suggestion
            const bool is_final_state = (particle.getGeneratorStatus() == 1);
```

**Comment by veprbl** [comment]
I took liberty to finish the transitions from vectors to collections, fixed some issues introduced while rebasing and reverted charged particle cut.

**Review by veprbl — APPROVED** [review]
Checked outputs before/after. This PR introduces only minimal changes: jets are placed into collections before the constituents.

**Comment by ruse-traveler** [comment]
@veprbl Thank you for taking care of those! Out of curiosity: is there a reason to revert the charged particle cut?

Either way, it's something we can discuss more when we revisit the constituent PODIO relations (i.e. for the next campaign). So this all looks good to me, and I'm okay with merging if everyone else is!

**Comment by veprbl** [comment]
Bundling changes places unnecessary burden on reviewers. Switch to multifactories should not cause any changes in the output, and it should be easy to demonstrate that. Thus, changes to the cuts don't belong in this change.

**Comment by veprbl** [comment]
@ruse-traveler Can you resubmit charge cut for review, or do you want me to do it? Please, create new branch in this repo this time. I also suggest renaming the collection to {Reconstructed,Generated}ChargedJets to communicate the difference in the result.

**Comment by ruse-traveler** [comment]
> @ruse-traveler Can you resubmit charge cut for review, or do you want me to do it? Please, create new branch in this repo this time. I also suggest renaming the collection to {Reconstructed,Generated}ChargedJets to communicate the difference in the result.

Can do! I'll put together a PR for that using a branch from this repo. And makes sense (re: reverting the cut)! I'll make sure to not sneak in extra changes in future PRs...
